src/nn/tensor_collection/collection.rs

```diff
@@ -74,6 +187,7 @@ impl<S: Shape, E: Dtype, D: DeviceStorage> TensorOptions<S, E, D> {
         TensorOptions {
             do_gradient_update: true,
             reset: |t| t.try_fill_with_zeros(),
+            shape: S::const_default(),
```

Comment:
Another option here is to only allow creating TensorOptions for `S: ConstShape`, which does impl Default. I'd prefer this than adding Default bounds to Shape & Dim.